test/Common/Include.hs

```diff
@@ -0,0 +1,94 @@
+-- This template expects CPP definitions for:
+--     PLATFORM_NAME = Posix | Windows
+--     IS_WINDOWS    = False | True
+
+{-# LANGUAGE TemplateHaskell #-}
+
+-- | Test functions that are common to Posix and Windows
+
+module Common.PLATFORM_NAME (extensionOperations) where
+
+import Data.Maybe (isJust, fromJust)
+import Control.Monad
+import Path.PLATFORM_NAME
+import System.FilePath.PLATFORM_NAME (pathSeparator)
+import Test.Hspec
+
+validExtensionsSpec :: String -> Path b File -> Path b File -> Spec
+validExtensionsSpec ext file fext = do
+    let f = show $ toFilePath file
+    let fx = show $ toFilePath fext
+
+    it ("addExtension " ++ show ext ++ " " ++ f ++ " == " ++ fx) $
+        addExtension ext file `shouldReturn` fext
+
+    it ("fileExtension " ++ fx ++ " == " ++ ext) $
+        fileExtension fext `shouldReturn` ext
+
+    it ("replaceExtension " ++ show ext ++ " " ++ fx ++ " == " ++ fx) $
+        replaceExtension ext fext `shouldReturn` fext
+
+extensionOperations :: String -> Spec
+extensionOperations rootDrive = do
+    let extension = ".foo"
+    let extensions = extension : [".foo.", ".foo.."]
+
+    describe "Only filenames and extensions" $
+      forM_ extensions $ \ext ->
+          forM_ filenames $ \f -> do
+              runTests parseRelFile f ext
+
+    describe "Relative dir paths" $
+      forM_ dirnames $ \d -> do
+          forM_ filenames $ \f -> do
+              let f1 = d ++ [pathSeparator] ++ f
+              runTests parseRelFile f1 extension
+
+    describe "Absolute dir paths" $
+      forM_ dirnames $ \d -> do
+          forM_ filenames $ \f -> do
+              let f1 = rootDrive ++ d ++ [pathSeparator] ++ f
+              runTests parseAbsFile f1 extension
+
+    -- Invalid extensions
+    forM_ invalidExtensions $ \ext -> do
+        it ("throws InvalidExtension when extension is [" ++ ext ++ "]")  $
+            addExtension ext $(mkRelFile "name")
+            `shouldThrow` (== InvalidExtension ext)
+
+    where
+
+    runTests parse file ext = do
+        let maybeFile = parse file
+        let maybeFileWithExt = parse (file ++ ext)
+        it ("Files " ++ file ++ " and " ++ (file ++ ext) ++ " are parsed successfully.") $
+            (isJust maybeFile, isJust maybeFileWithExt) `shouldBe` (True, True)
+        when (isJust maybeFile && isJust maybeFileWithExt) $
+            validExtensionsSpec ext (fromJust maybeFile) (fromJust maybeFileWithExt)
```

Comment:
I wrote this function to extract the common logic and added the part where we check if parsing went well -> otherwise, while pattern matching, the whole test suite would crash if parsing failed and tests after these tests would never run.

Reply:
That's nice, but please pattern-match instead of using `when`, `isJust` and `fromJust`.

You can pattern-match on both using `case (maybeFile, maybeFileWithExt)` if necessary.

Reply:
Sure, good point!

I modified this part to use pattern matching, which also removed some of the duplication (double checking if both are Just).

The main part I am somewhat struggling with though is creating this intentionally failing test.
Basically, what I want is to report a failure if parsing went wrong (and it should offer useful information) and skip the tests that can't be performed because of it. If parsing went right, I want to continue with those tests.
The way I currently report this failure is by constructing this additional test that I know will fail. I tried to find a way to construct it explicitly (smth like failTest "information"), but it seems HSpec offers no such mechanism (at least I could not find one), so I did this slight abuse of `shouldBe`.

So while I think this achieves the purpose in practice (failed test is reported with clear message, and other tests continue), I am not perfectly happy with `shouldBe "(Just _, Just _)"` part, and I am not perfectly happy with the fact that this causes number of tests to vary (since we now have branching), but I don't currently see any way around it.